lib/Support/TestPasses.cpp

```diff
@@ -0,0 +1,112 @@
+//===- TestPasses.cpp - Test passes for the support infrastructure -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements test passes for the support infrastructure.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/ESIOps.h"
+#include "circt/Support/CPPCDE.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace cppcde;
+
+//===----------------------------------------------------------------------===//
+// CPPCDE passes.
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct TestCPPCDEPass
+    : public PassWrapper<TestCPPCDEPass, OperationPass<mlir::ModuleOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestCPPCDEPass)
+
+  void runOnOperation() override;
+  StringRef getArgument() const override { return "test-cppcde"; }
+  StringRef getDescription() const override {
+    return "Runs various built-in cppcde tests.";
+  }
+
+  void getDependentDialects(::mlir::DialectRegistry &registry) const override {
+    registry.insert<hw::HWDialect, comb::CombDialect, seq::SeqDialect,
+                    esi::ESIDialect>();
+  }
+};
+} // namespace
+
+class MyAdder : public HWModuleGenerator<DefaultCDEValue> {
+public:
+  using HWModuleGenerator::HWModuleGenerator;
+  std::string getName() override { return "MyAdder"; }
+  void createIO() override {
+    input("in0", ctx.b.getIntegerType(32));
+    input("in1", ctx.b.getIntegerType(32));
+    clock();
+    output("out0", ctx.b.getIntegerType(32));
+  }
+
+  void generate(CDEPorts &ports) override {
+    ports["out0"].assign((ports["in0"] + ports["in1"]).reg("add_reg"));
+  }
+};
+
+// Example of specializing the generator with a custom value type that extends
+// the default value type with a few extra operators.
+class MyESIAdder : public HWModuleGenerator<ESICDEValue> {
+public:
+  using HWModuleGenerator::HWModuleGenerator;
+  std::string getName() override { return "MyESIAdder"; }
+  void createIO() override {
+    Type channelType =
+        esi::ChannelType::get(ctx.b.getContext(), ctx.b.getIntegerType(32));
```

Comment:
In the future, we should consider adding convenience type constructors as well, so this could be something like `createType<esi::ChannelType>(32)`.